### PR TITLE
Add a qa engineer to write tests and run tests

### PR DESCRIPTION
**Pull request by garylin2099**
Update:
1. implement three new actions: WriteTest, RunCode, DebugError
2. implement QaEngineer, who will write tests and run tests to check Engineer's code from upstream. It will analyze the testing results, direct error messages, if any, to the appropriate coder (Engineer or itself). It will then rewrite problematic parts of its test codes once receiving the errors. This consists of a "RunCode -> DebugError -> RunCode -> DebugError -> ..." loop, which makes the code more runnable.
3. A first bite of the natural communication between agents. Half implement a RunCode result msg sent from QaEngineer to Engineer.
4. Try ```python startup.py 'Write a flappy bird game' --n_round 10``` or ```python startup.py 'Write a black jack game' --n_round 10``` for a demo. A good case is that QaEngineer found Engineer missed some import statements in its code and instructed it to fix.

TODO:
1. "RunCode -> DebugError" loop is currently only applied to QaEngineer, if the implementation looks fine, we may extend this form to Engineer as well. In this way, Engineer can handle errors sent to it from QaEngineer, rewrite the code and send it back to QaEngineer for another tests. This allows meaningful debugging on the development codes.
2. To allow communication between agents, I add two attributes for a Message, sent_from and send_to. However, I hard code the sender and recipient when making a message in a role run. We need a better way.
3. For a role capable of multiple actions, we may place the action choosing, either human-defined or LLM-self-generated, under the _think() function for better compliance with the framework design. I currently placed it under _act() for the QaEngineer, out of convenience.
4. We could find a better name for the Memory.remember() function? The function is used to find dissimilar msg whereas "remember" sounds like finding the similar ones. This is a bit confusing.

Problems:
1. Difficulties testing snake game where most logic involves ```curses``` package, which is hard to mock. Black jack and flappy bird are better, where the test codes find real bugs to be solved.

### CODE REVIEW

File: metagpt/actions/write_test.py [hunk 1]
```diff
@@ -5,22 +5,45 @@
...
 from metagpt.actions.action import Action
+from metagpt.utils.common import CodeParser
 
+PROMPT_TEMPLATE = """
+NOTICE
+1. Role: You are a QA engineer; the main goal is to design, develop, and execute PEP8 compliant, well-structured, maintainable test cases and scripts for Python 3.9. Your focus should be on ensuring the product quality of the entire project through systematic testing.
+2. Requirement: Based on the context, develop a comprehensive test suite that adequately covers all relevant aspects of the code file under review. Your test suite will be part of the overall project QA, so please develop complete, robust, and reusable test cases.
+3. Attention1: Use '##' to split sections, not '#', and '## <SECTION_NAME>' SHOULD WRITE BEFORE the test case or script.
+4. Attention2: If there are any settings in your tests, ALWAYS SET A DEFAULT VALUE, ALWAYS USE STRONG TYPE AND EXPLICIT VARIABLE.
+5. Attention3: YOU MUST FOLLOW "Data structures and interface definitions". DO NOT CHANGE ANY DESIGN. Make sure your tests respect the existing design and ensure their validity.
```
Comment (better629) [inline metagpt/actions/write_test.py:18]: their validity -> its validity

File: metagpt/actions/write_test.py [hunk 2]
```diff
@@ -5,22 +5,45 @@
...
+6. Think before writing: What should be tested and validated in this document? What edge cases could exist? What might fail?
+7. CAREFULLY CHECK THAT YOU DON'T MISS ANY NECESSARY TEST CASES/SCRIPTS IN THIS FILE.
+Attention: Use '##' to split sections, not '#', and '## <SECTION_NAME>' SHOULD WRITE BEFORE the test case or script and triple quote.
+-----
+## Given the following code, please write appropriate test cases using Python's unittest framework to verify the correctness and robustness of this code:
```
Comment (better629) [inline metagpt/actions/write_test.py:23]: suggestion: pytest is more easy

File: metagpt/actions/write_test.py [hunk 3]
```diff
@@ -5,22 +5,45 @@
...
+```python
+{code_to_test}
+```
+Note that the code to test is at {source_file_path}, we will put your test code at {workspace}/tests/{test_file_name}, and run your test code from {workspace},
+you should correctly import the necessary classes based on these file locations!
+## {test_file_name}: Write test code with triple quoto. Do your best to implement THIS ONLY ONE FILE.
```
Comment (better629) [inline metagpt/actions/write_test.py:29]: triple quoto -> triple quotes?

File: metagpt/roles/qa_engineer.py [hunk 4]
```diff
@@ -5,11 +5,158 @@
...
+from metagpt.logs import logger
+from metagpt.roles import Role
+from metagpt.schema import Message
+from metagpt.roles.engineer import Engineer
+from metagpt.utils.common import CodeParser, parse_recipient
+from metagpt.utils.special_tokens import MSG_SEP, FILENAME_CODE_SEP
 
 class QaEngineer(Role):
-    def __init__(self, name, profile, goal, constraints):
+    def __init__(self, name="Edward", profile="QA Engineer",
```
Comment (better629) [inline metagpt/roles/qa_engineer.py:22]: How about keep `profile` and `send_to` with same value?

File: metagpt/roles/qa_engineer.py [hunk 5]
```diff
@@ -5,11 +5,158 @@
...
+                content=str(file_info), role=self.profile, cause_by=WriteTest,
+                sent_from="QaEngineer", send_to="QaEngineer"
+            )
+            self._publish_message(msg)
+        
+        logger.info(f'Done {self.get_workspace()}/tests generating.')
+    
+    async def _run_code(self, msg):
+        file_info = eval(msg.content)
+        code_to_test = open(file_info["file_path"], "r").read()
```
Comment (better629) [inline metagpt/roles/qa_engineer.py:97]: maybe it's better to add a file if-exist check

File: metagpt/roles/qa_engineer.py [hunk 6]
```diff
@@ -5,11 +5,158 @@
...
+        if file_name:
+            self.write_file(file_name, code)
+            recipient = msg.sent_from # send back to the one who ran the code for another run, might be one's self
+            msg = Message(content=file_info, role=self.profile, cause_by=DebugError, sent_from="QaEngineer", send_to=recipient)
+            self._publish_message(msg)
+    
+    async def _observe(self) -> int:
+        await super()._observe()
+        self._rc.news = [msg for msg in self._rc.news \
+            if msg.send_to == "QaEngineer"] # only relevant msgs count as observed news
```
Comment (better629) [inline metagpt/roles/qa_engineer.py:134]: A role only deal the msgs which relevant to the `_watch`. How about use this strategy to filter?

File: metagpt/roles/role.py [hunk 7]
```diff
@@ -70,6 +70,7 @@ class RoleContext(BaseModel):
     state: int = Field(default=0)
     todo: Action = Field(default=None)
     watch: set[Type[Action]] = Field(default_factory=set)
+    news: list[Type[Message]] = Field(default=[])
```
Comment (better629) [inline metagpt/roles/role.py:73]: a little repetition with `important_memory`, no need to add.

File: tests/metagpt/actions/test_debug_error.py [hunk 8]
```diff
@@ -17,7 +151,6 @@ async def test_debug_error():
 
     debug_error = DebugError("debug_error")
 
-    result = await debug_error.run(code, error)
+    file_name, rewritten_code = await debug_error.run(context=EXAMPLE_MSG_CONTENT)
 
-    # mock_llm.ask.assert_called_once_with(prompt)
-    assert len(result) > 0
+    assert "class TestPlayer" in rewritten_code
```
Comment (better629) [inline tests/metagpt/actions/test_debug_error.py:156]: case failed with
================================================================== short test summary info ===================================================================
FAILED tests/metagpt/actions/test_debug_error.py::test_debug_error - assert 'class TestPlayer' in 'from typing import List\nfrom deck import Deck\nfrom card import Card\n\nclass Player:\n    """\n    A class represen...   ...

File: metagpt/actions/run_code.py [hunk 9]
```diff
@@ -6,20 +6,117 @@
...
+            outs, errs = await self.run_text(code=code)
+        
+        logger.info(f"{outs=}")
+        logger.info(f"{errs=}")
+
+        context = CONTEXT.format(
+            code=code, code_file_name=code_file_name,
+            test_code=test_code, test_file_name=test_file_name,
+            command=" ".join(command),
+            outs=outs[:500], # outs might be long but they are not important, truncate them to avoid token overflow
```
Comment (better629) [inline metagpt/actions/run_code.py:113]: what if outs is None ?

File: tests/metagpt/actions/test_run_code.py [hunk 10]
```diff
@@ -6,33 +6,65 @@
...
+    result = await action.run(
+        mode="script",
+        code='python -c "print(1/0)"',
+        code_file_name="",
+        test_code="",
+        test_file_name="",
+        command=["python", "-c", "print(1/0)"],
+        working_directory=".",
+        additional_python_paths=[]
+    )
```
Comment (better629) [inline tests/metagpt/actions/test_run_code.py:69]: test case fails with
================================================================== short test summary info ===================================================================
FAILED tests/metagpt/actions/test_run_code.py::test_run - TypeError: 'NoneType' object is not subscriptable

**Comment by better629** [comment]
[blackjack_game_with_qa.zip](https://github.com/geekan/MetaGPT/files/12246388/blackjack_game_with_qa.zip)
[blackjack_game_without_qa.zip](https://github.com/geekan/MetaGPT/files/12246395/blackjack_game_without_qa.zip)

`blackjack_game_with_qa` 
image

After click `instructions`, it occurs with  
```
    tk.messagebox.showinfo("Instructions", instructions)
AttributeError: module 'tkinter' has no attribute 'messagebox'
```

`blackjack_game_without_qa` 
It can star the server, but also occurs with resouce not found error like
```
    raise TemplateNotFound(template)
jinja2.exceptions.TemplateNotFound: index.html
```

Maybe you should given a fail-to-success case. and give a larger `n_round` in example. 10 is small and it will exit before `Engineer` rewrite code.

Further, do you have consider the pkg version problem in `RunCode` and how to instruct `Engineer` to solve it?

File: metagpt/actions/write_test.py (hunk 2 above)
Comment (garylin2099) [reply metagpt/actions/write_test.py:23]: 但是要装包，或者需要跑mg的环境有pytest？

**Comment by better629** [comment]
The answer to `We could find a better name for the Memory.remember() function? `
`remember` is not only for similar memories(long-term) but also clear memory(short-term) as comment mentioned.

File: metagpt/roles/qa_engineer.py (hunk 6 above)
Comment (garylin2099) [reply metagpt/roles/qa_engineer.py:134]: 用_watch会有问题，因为不同role可能watch同一个action，无法区分到底谁来处理执行这个action以避免重复，应该要允许engineer和qa_engineer都可以debug

File: metagpt/roles/role.py (hunk 7 above)
Comment (garylin2099) [reply metagpt/roles/role.py:73]: 仅用important memory无法区分哪些message处理过了，哪些未处理，最好就是在每次observe的时候，把news存下来

File: metagpt/actions/run_code.py (hunk 9 above)
Comment (garylin2099) [reply metagpt/actions/run_code.py:113]: outs为空会是一个空字符""，不会是None的

**Comment by garylin2099** [comment]
> [blackjack_game_with_qa.zip](https://github.com/geekan/MetaGPT/files/12246388/blackjack_game_with_qa.zip) [blackjack_game_without_qa.zip](https://github.com/geekan/MetaGPT/files/12246395/blackjack_game_without_qa.zip)
> 
> `blackjack_game_with_qa` image
> 
> After click `instructions`, it occurs with
> 
> ```
>     tk.messagebox.showinfo("Instructions", instructions)
> AttributeError: module 'tkinter' has no attribute 'messagebox'
> ```
> 
> `blackjack_game_without_qa` It can star the server, but also occurs with resouce not found error like
>  
> ```
>     raise TemplateNotFound(template)
> jinja2.exceptions.TemplateNotFound: index.html
> ```
> 
> Maybe you should given a fail-to-success case. and give a larger `n_round` in example. 10 is small and it will exit before `Engineer` rewrite code.
> 
> Further, do you have consider the pkg version problem in `RunCode` and how to instruct `Engineer` to solve it?

1. I currently only apply DebugError to QaEngineer, to avoid making too many changes on the existing Engineer, so Engineer can't rewrite its code and the bug in the development code remains there unchanged. I'm thinking of apply DebugError action to Engineer in a subsequent PR if the current implementation looks fine.
2. Yes but still thinking, installing pkg might be a harder question, we can solve native python code problem first

File: metagpt/actions/run_code.py [hunk 11]
```diff
@@ -6,20 +6,117 @@
...
+"""
 
 class RunCode(Action):
-    def __init__(self, name, context=None, llm=None):
+    def __init__(self, name="RunCode", context=None, llm=None):
         super().__init__(name, context, llm)
 
-    async def run(self, code):
+    @classmethod
+    async def run_text(cls, code):
```
[the remark this answers is not in the document]
Comment (garylin2099) [reply metagpt/actions/run_code.py]: done

File: tests/metagpt/actions/test_debug_error.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (garylin2099) [reply tests/metagpt/actions/test_debug_error.py]: done

File: metagpt/actions/debug_error.py [hunk 12]
```diff
@@ -5,15 +5,47 @@
...
+1. Role: You are a Development Engineer or QA engineer;
+2. Task: You received this message from another Development Engineer or QA engineer who run or test your code. 
+Based on the message, first, figure out your own role, i.e. Engineer or QaEngineer,
+then rewrite the development code or the test code based on your role, the error, and the summary, such that all bugs are fixed and the code performs well.
+Attention: Use '##' to split sections, not '#', and '## <SECTION_NAME>' SHOULD WRITE BEFORE the test case or script and triple quote.
+The message is as follows:
+{context}
+---
+Now you should start rewriting the code:
+## file name of the code to rewrite: Write code with triple quoto. Do your best to implement THIS ONLY ONE FILE.
```
[the remark this answers is not in the document]
Comment (garylin2099) [reply metagpt/actions/debug_error.py]: done

File: metagpt/actions/debug_error.py [hunk 13]
```diff
@@ -5,15 +5,47 @@
...
+import re
 
+from metagpt.logs import logger
+from metagpt.actions.action import Action
+from metagpt.utils.common import CodeParser
 
+PROMPT_TEMPLATE = """
+NOTICE
+1. Role: You are a Development Engineer or QA engineer;
+2. Task: You received this message from another Development Engineer or QA engineer who run or test your code.
```
[the remark this answers is not in the document]
Comment (garylin2099) [reply metagpt/actions/debug_error.py]: done

File: metagpt/actions/debug_error.py [hunk 14]
```diff
@@ -5,15 +5,47 @@
...
+from metagpt.actions.action import Action
+from metagpt.utils.common import CodeParser
 
+PROMPT_TEMPLATE = """
+NOTICE
+1. Role: You are a Development Engineer or QA engineer;
+2. Task: You received this message from another Development Engineer or QA engineer who run or test your code. 
+Based on the message, first, figure out your own role, i.e. Engineer or QaEngineer,
+then rewrite the development code or the test code based on your role, the error, and the summary, such that all bugs are fixed and the code performs well.
+Attention: Use '##' to split sections, not '#', and '## <SECTION_NAME>' SHOULD WRITE BEFORE the test case or script and triple quote.
```
[the remark this answers is not in the document]
Comment (garylin2099) [reply metagpt/actions/debug_error.py]: done

File: metagpt/actions/write_test.py (hunk 1 above)
Comment (garylin2099) [reply metagpt/actions/write_test.py]: done

File: metagpt/actions/write_test.py (hunk 3 above)
Comment (garylin2099) [reply metagpt/actions/write_test.py:29]: done

File: metagpt/roles/qa_engineer.py (hunk 4 above)
Comment (garylin2099) [reply metagpt/roles/qa_engineer.py]: done

File: metagpt/roles/qa_engineer.py (hunk 5 above)
Comment (garylin2099) [reply metagpt/roles/qa_engineer.py]: done

File: metagpt/actions/run_code.py (hunk 9 above)
Comment (garylin2099) [reply metagpt/actions/run_code.py:114]: done

File: tests/metagpt/actions/test_run_code.py (hunk 10 above)
Comment (garylin2099) [reply tests/metagpt/actions/test_run_code.py:69]: done

File: metagpt/actions/run_code.py (hunk 9 above)
Comment (better629) [reply metagpt/actions/run_code.py:114]: need a type check if outs is None

Comment (garylin2099) [reply metagpt/actions/run_code.py:114]: 我在输出outs的内部函数里限定outs不会是None了

File: metagpt/actions/write_test.py (hunk 2 above)
Comment (stellaHSR) [reply metagpt/actions/write_test.py:23]: how about ask LLM to generate testsuites via pytest/unittest? and just run 'pip install pytest' cmd in run_code func.

File: metagpt/roles/qa_engineer.py [hunk 15]
```diff
@@ -86,22 +87,27 @@ async def _write_test(self, message: Message) -> None:
             }
             msg = Message(
```
Comment (stellaHSR) [inline metagpt/roles/qa_engineer.py:88]: in this way, the testcase looks like unittests

File: metagpt/actions/debug_error.py [hunk 16]
```diff
@@ -14,15 +14,15 @@
...
 NOTICE
 1. Role: You are a Development Engineer or QA engineer;
-2. Task: You received this message from another Development Engineer or QA engineer who run or test your code. 
+2. Task: You received this message from another Development Engineer or QA engineer who ran or tested your code. 
 Based on the message, first, figure out your own role, i.e. Engineer or QaEngineer,
 then rewrite the development code or the test code based on your role, the error, and the summary, such that all bugs are fixed and the code performs well.
-Attention: Use '##' to split sections, not '#', and '## <SECTION_NAME>' SHOULD WRITE BEFORE the test case or script and triple quote.
+Attention: Use '##' to split sections, not '#', and '## <SECTION_NAME>' SHOULD WRITE BEFORE the test case or script and triple quotes.
 The message is as follows:
 {context}
```
Comment (stellaHSR) [inline metagpt/actions/debug_error.py:22]: add code and test code is better

File: metagpt/actions/write_test.py (hunk 2 above)
Comment (garylin2099) [reply metagpt/actions/write_test.py:23]: 我觉得这可以列为一个需求点，即在runcode过程中实现装包，放新的PR里实现

File: metagpt/actions/run_code.py [hunk 17]
```diff
@@ -6,20 +6,57 @@
...
+        working_directory = str(working_directory)
+        additional_python_paths = [str(path) for path in additional_python_paths]
+
+        # Copy the current environment variables
+        env = os.environ.copy()
+
+        # Modify the PYTHONPATH environment variable
+        additional_python_paths = [working_directory] + additional_python_paths
+        additional_python_paths = ":".join(additional_python_paths)
+        env['PYTHONPATH'] = additional_python_paths + ':' + env.get('PYTHONPATH', '')
```
[the remark this answers is not in the document]
Comment (garylin2099) [reply metagpt/actions/run_code.py:84]: 是的

File: metagpt/roles/qa_engineer.py [hunk 18]
```diff
@@ -5,11 +5,93 @@
...
+    async def _act(self) -> Message:
+        code_action_watched = self._rc.important_memory[-1]
+        code_msgs = code_action_watched.content.split(WRITECODE_MSG_SEP)
+        for code_msg in code_msgs:
+            
+            # write tests
+            file_name, file_path, code_to_test = code_msg.split(FILENAME_CODE_SEP)
+            test_file_name = "test_" + file_name
+            logger.info(f'Writing {test_file_name}..')
+            code = await WriteTest().run(
```
[the remark this answers is not in the document]
Comment (garylin2099) [reply metagpt/roles/qa_engineer.py]: 对，针对每个文件进行单元测试。谁来写这里有一个细节问题，engineer的workspace对应的是 pkg/pkg ，要让它写到 pkg/tests 下有些麻烦，所以现在是设计 pkg/pkg 下的代码，也就是开发的代码都由engineer来写，测试的代码都由qa_engineer来写

File: metagpt/actions/run_code.py [hunk 19]
```diff
@@ -6,20 +6,57 @@
...
+    @classmethod
+    async def run_script(cls, working_directory, additional_python_paths=[], command=[]):
+        working_directory = str(working_directory)
+        additional_python_paths = [str(path) for path in additional_python_paths]
+
+        # Copy the current environment variables
+        env = os.environ.copy()
+
+        # Modify the PYTHONPATH environment variable
+        additional_python_paths = [working_directory] + additional_python_paths
```
[the remark this answers is not in the document]
Comment (garylin2099) [reply metagpt/actions/run_code.py:82]: 涉及装包的，我觉得下一个优化实现吧

File: metagpt/roles/qa_engineer.py (hunk 15 above)
Comment (garylin2099) [reply metagpt/roles/qa_engineer.py:88]: 是的。单元测试只需要考虑一个文件，把其他逻辑隔离，我觉得挺适合的；此外，这种交互型的游戏，还没想到怎么比较好地通过脚本进行集成测试

File: metagpt/actions/debug_error.py (hunk 16 above)
Comment (garylin2099) [reply metagpt/actions/debug_error.py:22]: 写code和写test code分开的理由见上

File: metagpt/roles/qa_engineer.py [hunk 20]
```diff
@@ -5,11 +5,93 @@
...
+        return WORKSPACE_ROOT / workspace / workspace
+        
+    
+    def write_file(self, filename: str, code: str):
+        workspace = self.get_workspace() / 'tests'
+        file = workspace / filename
+        file.parent.mkdir(parents=True, exist_ok=True)
+        file.write_text(code)
+
+    def recv(self, message: Message) -> None:
```
[the remark this answers is not in the document]
Comment (garylin2099) [reply metagpt/roles/qa_engineer.py:55]: done

**Comment by better629** [comment]
LGTM so far, and look forward to see the `rewrite` feature.

**Comment by stellaHSR** [comment]
> LGTM so far, and look forward to see the `rewrite` feature.

LGTM.
